hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/datanode/BPOfferService.java

```diff
@@ -1163,14 +1105,60 @@ void notifyNamenodeDeletedBlockInt(
       // In case of un/successful block reports we have to inform the leader that
       // block reporting has finished for now.
       if (blkReportHander != null) {
-        getLeaderActor().blockReportCompleted(bpRegistration);
+        BPServiceActor leader =  getLeaderActor();
+        if( leader != null){
+          leader.blockReportCompleted(bpRegistration);
+        }
       }
     }
 
     scheduleNextBlockReport(startTime);
     return cmds.size() == 0 ? null : cmds;
   }
 
+  StorageBlockReport[] slimBlockReports(StorageBlockReport[] reports){
+    StorageBlockReport[] slimStorageReports = new StorageBlockReport[reports.length];
+
+    for(int i = 0 ; i < reports.length; i++){
+      StorageBlockReport fatSR = reports[i];
+      BlockReport fatBR = fatSR.getReport();
+      Bucket slimBuckets[] = new Bucket[fatBR.getBuckets().length];
+
+      for(int j = 0; j < fatBR.getBuckets().length; j++){
+        Bucket slimBucket = new Bucket();
+        slimBucket.setHash(fatBR.getBuckets()[j].getHash());
+        slimBucket.setBlocks(BlockListAsLongs.EMPTY);
+        slimBuckets[j] = slimBucket;
+      }
+
+      BlockReport slimBR = new BlockReport(slimBuckets, fatBR.getNumberOfBlocks());
+
+      slimStorageReports[i] = new StorageBlockReport(fatSR.getStorage(), slimBR);
+    }
+
+    return slimStorageReports;
+  }
+
+  public void removeMatchingBuckets(DatanodeCommand cmd, StorageBlockReport reports[]){
+    HashesMismatchCommand hmcmd = (HashesMismatchCommand) cmd;
+    Map<String, List<Integer>> map = ((HashesMismatchCommand) cmd).getMissMatchingBuckets();
+    for(StorageBlockReport report : reports){
+      List<Integer> mismatchingBuckets = map.get(report.getStorage().getStorageID());
+      if(mismatchingBuckets != null && mismatchingBuckets.size()>0){
+        removeMatchingBuckets(mismatchingBuckets,report.getReport());
+      }
+    }
+  }
+
+  public static void removeMatchingBuckets(List<Integer> mismatchingBuckets, BlockReport report){
+    for(int i = 0; i < report.getBuckets().length; i++){
+      if(!mismatchingBuckets.contains(i)){
```

Comment:
contains on a list is not efficient, the type of mismatchingBuckets should probably be a set.

Reply:
mismatchingBuckets is expected to contain only few elements. Only at cluster restart it will have at most 1000 elements. Set will not help, but HashSet will.